waspc/src/Wasp/Generator/Monad.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Wasp.Generator.Monad
+  ( Generator,
+    GeneratorError (..),
+    GeneratorWarning (..),
+    catchGeneratorError,
+    logAndThrowGeneratorError,
+    logGeneratorWarning,
+    runGenerator,
+  )
+where
+
+import Control.Arrow (left)
+import Control.Monad.Except (ExceptT, MonadError (throwError), runExceptT)
+import qualified Control.Monad.Except as MonadExcept
+import Control.Monad.Identity (Identity (runIdentity))
+import Control.Monad.State (MonadState, StateT (runStateT), modify)
+import Data.List.NonEmpty (NonEmpty, fromList)
+
+-- | Generator is a monad transformer stack where we abstract away the underlying
+-- concrete monad transformers with the helper functions below. This will allow us
+-- to refactor and add more transformers (or swap them) without any caller changes.
+--
+-- The outer Either layer represents the error that halted generation. Every error is fatal.
+-- There may optionally be additional errors or non-fatal warnings logged in the State.
+newtype Generator a = Generator
+  { _runGenerator :: ExceptT GeneratorError (StateT GeneratorState Identity) a
+  }
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadState GeneratorState,
+      MonadError GeneratorError
+    )
+
+data GeneratorState = GeneratorState
+  { warnings :: [GeneratorWarning],
+    errors :: [GeneratorError]
+  }
+
+data GeneratorError = GenericGeneratorError String
+
+instance Show GeneratorError where
+  show (GenericGeneratorError e) = e
+
+data GeneratorWarning = GenericGeneratorWarning String
+
+instance Show GeneratorWarning where
+  show (GenericGeneratorWarning e) = e
+
+-- Runs the generator and either returns a result, or a list of 1 or more errors.
+-- Even if successful there may be warnings, so they are always included.
```

Comment:
Nice comment, I think is a very clear explanation!

Reply:
And what I love about Haskell is that even if this comment was not here, you could get all of this information just from the type! In theory, this comment is redundant. Although I think it is nice to have it here since it explains it in more human way.

Reply:
Agreed! Went from comments telling a story to the type system, but agree it is nice to have it reiterated in plain English here as well.